File: FE_vpn/src/api/machines.js
import { API_BASE_URL, request } from './client'

function buildQuery(params = {}) {
    const entries = Object.entries(params)
        .filter(([, value]) => value !== undefined && value !== null && value !== '')
        .map(([key, value]) => [key, String(value)])
    const search = new URLSearchParams(entries)
    const query = search.toString()
    return query ? `?${query}` : ''
}

export async function listMachines(params, token) {
    const query = buildQuery(params)
    return request(`/machines${query}`, { token })
}

export async function getMachine(machineId, token) {
    return request(`/machines/${machineId}`, { token })
}

export async function startMachine(machineId, token) {
    return request(`/machines/${machineId}/start`, { method: 'POST', token })
}

export async function resumeMachine(machineId, token) {
    return request(`/machines/${machineId}/resume`, { method: 'POST', token })
}

export async function getActiveSession(token) {
    return request('/machines/sessions/active', { token })
}

export async function getSessionHistory(params = {}, token) {
    const query = buildQuery({
        page: params.page,
        page_size: params.pageSize,
        status: params.status,
        machine_id: params.machineId,
        date_from: params.dateFrom,
        date_to: params.dateTo,
        sort: params.sort,
    })
    return request(`/machines/sessions/history${query}`, { token })
}

export async function getSessionHistorySummary(token) {
    return request('/machines/sessions/history/summary', { token })
}

export async function stopSession(sessionId, token, options = {}) {
    const query = buildQuery({ retain_snapshot: options.retainSnapshot ? true : undefined })
    return request(`/machines/sessions/${sessionId}/stop${query}`, { method: 'POST', token })
}

export async function heartbeatSession(sessionId, payload = {}, token) {
    return request(`/machines/sessions/${sessionId}/heartbeat`, { method: 'POST', body: payload, token })
}

export async function checkVpnConnection(sessionId, token) {
    return request(`/machines/sessions/${sessionId}/vpn/check`, { method: 'POST', token })
}

export async function markSunshinePaired(sessionId, token) {
    return request(`/machines/sessions/${sessionId}/sunshine/pair`, { method: 'POST', token })
}

function parseDownloadFilename(disposition, fallback) {
    const match = disposition?.match(/filename="?([^"]+)"?/i)
    return match?.[1] || fallback
}

export async function downloadOvpn(sessionId, token) {
    const headers = {}
    if (token) headers.Authorization = `Bearer ${token}`

    const path = `/machines/sessions/${sessionId}/ovpn`
    const url = API_BASE_URL ? `${API_BASE_URL}${path}` : path
    const response = await fetch(url, { headers })

    if (!response.ok) {
        const text = await response.text()
        let data = null
        try {
            data = text ? JSON.parse(text) : null
        } catch {
            data = text
        }
        const message = data?.detail || data?.message || `Request failed (${response.status})`
        throw new Error(message)
    }

    const blob = await response.blob()
    const filename = parseDownloadFilename(
        response.headers.get('Content-Disposition'),
        `vpngaming-${String(sessionId).slice(0, 8)}.ovpn`,
    )
    return { blob, filename }
}

File: BE_vpn/app/api/machines.py
from uuid import UUID
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from fastapi.responses import Response
from sqlalchemy.orm import Session

from app import models, schemas
from app.api.deps import get_current_user, get_optional_current_user
from app.database import get_db
from app.services.machine_service import MachineService


router = APIRouter(prefix="/machines", tags=["machines"])


def get_machine_service(db: Session = Depends(get_db)) -> MachineService:
    return MachineService(db)


@router.get("", response_model=schemas.MachinesPage)
def list_machines(
    page: int = Query(1, ge=1),
    page_size: int = Query(12, ge=1, le=100),
    region: str | None = Query(None),
    gpu: str | None = Query(None),
    status_filter: str | None = Query(None, alias="status"),
    min_ping: int | None = Query(None, ge=0),
    max_ping: int | None = Query(None, ge=0),
    sort: str = Query("best", pattern="^(best|ping)$"),
    current_user: models.User | None = Depends(get_optional_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.list_machines(
        page=page,
        page_size=page_size,
        region=region,
        gpu=gpu,
        status_filter=status_filter,
        min_ping=min_ping,
        max_ping=max_ping,
        sort=sort,
        current_user=current_user,
    )


@router.get("/sessions/active", response_model=schemas.SessionOut | None)
def get_active_user_session(
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.get_active_user_session(current_user)


@router.get("/sessions/history", response_model=schemas.SessionHistoryPage)
def get_user_session_history(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=50),
    status_filter: str | None = Query(None, alias="status"),
    machine_id: UUID | None = Query(None),
    date_from: datetime | None = Query(None),
    date_to: datetime | None = Query(None),
    sort: str = Query("recent", pattern="^(recent|oldest)$"),
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.get_user_session_history(
        current_user=current_user,
        page=page,
        page_size=page_size,
        status_filter=status_filter,
        machine_id=machine_id,
        date_from=date_from,
        date_to=date_to,
        sort=sort,
    )


@router.get("/sessions/history/summary", response_model=schemas.SessionHistorySummaryOut)
def get_user_session_history_summary(
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.get_user_session_history_summary(current_user)


@router.post("/sessions/{session_id}/stop", response_model=schemas.SessionOut)
def stop_user_session(
    session_id: UUID,
    retain_snapshot: bool = Query(False),
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.stop_user_session(session_id, current_user, retain_snapshot=retain_snapshot)


@router.post("/sessions/{session_id}/heartbeat", response_model=schemas.SessionOut)
def heartbeat_session(
    session_id: UUID,
    payload: schemas.SessionHeartbeatRequest,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.record_session_heartbeat(session_id, current_user, payload)


@router.get("/sessions/{session_id}/ovpn")
def download_session_ovpn(
    session_id: UUID,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    filename, content = machine_service.build_session_ovpn(session_id, current_user)
    return Response(
        content=content,
        media_type="application/x-openvpn-profile",
        headers={
            "Content-Disposition": f'attachment; filename="{filename}"',
            "Cache-Control": "no-store",
            "X-Content-Type-Options": "nosniff",
        },
    )


@router.post("/sessions/{session_id}/vpn/check", response_model=schemas.SessionOut)
def check_session_vpn(
    session_id: UUID,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.verify_vpn_connection(session_id, current_user)


@router.post("/sessions/{session_id}/sunshine/pair", response_model=schemas.SessionOut)
def pair_session_sunshine(
    session_id: UUID,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.mark_sunshine_paired(session_id, current_user)


@router.get("/{machine_id}", response_model=schemas.MachineDetailOut)
def get_machine_detail(
    machine_id: UUID,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.get_machine_detail(machine_id, current_user)


@router.post("/{machine_id}/start", response_model=schemas.SessionOut)
def start_machine_session(
    machine_id: UUID,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.start_machine_session(machine_id, current_user)


@router.post("/{machine_id}/resume", response_model=schemas.SessionOut)
def resume_machine_session(
    machine_id: UUID,
    current_user: models.User = Depends(get_current_user),
    machine_service: MachineService = Depends(get_machine_service),
):
    return machine_service.resume_machine_session(machine_id, current_user)
